src/helpers/checks.jl

```diff
@@ -48,57 +124,288 @@ function check_differential(
+        )
+            return false
+        end
+    end
+    if mode === :CriticalPoint # find a critical point and update grad, hess and X
+        p = gradient_descent(M, f, grad_f, p)
+        gradient = grad_f(M, p)
+        X = rand(M; vector_at=p)
+        Hessian = Hess_f(M, p, X)
+    end
+    #
+    # slope check
+    X_n = X ./ norm(M, p, X) # normalize tangent direction
+    Hessian_n = Hessian ./ norm(M, p, X)
+    # function for the directional derivative
+    #
+    T = exp10.(log_range)
+    # points p_i to evaluate our error function at
+    points = map(t -> retract(M, p, X_n, t, retraction_method), T)
+    # F(p_i)
+    costs = [f(M, pi) for pi in points]
+    # linearized
+    linearized = map(
+        t ->
+            f(M, p) +
+            t * real(inner(M, p, gradient, X_n)) +
+            t^2 / 2 * real(inner(M, p, Hessian_n, X_n)),
+        T,
+    )
+    return prepare_check_result(
+        log_range,
+        abs.(costs .- linearized),
+        3.0;
+        exactness_tol=exactness_tol,
+        io=io,
+        name="Hessian",
+        plot=plot,
+        slope_tol=slope_tol,
+        throw_error=throw_error,
+        window=window,
+    )
+end
+
+@doc raw"""
+    is_Hessian_linear(M, Hess_f, p,
+    X=rand(M; vector_at=p), Y=rand(M; vector_at=p), a=randn(), b=randn();
+    throw_error=false, io=nothing, kwargs...
+)
+
+Check whether the Hessian function `Hess_f` fulfills linearity, i.e. that
+
+```math
+\operatorname{Hess} f(p)[aX + bY] = b\operatorname{Hess} f(p)[X]
+ + b\operatorname{Hess} f(p)[Y]
+```
+
+which is checked using `isapptox` and the `kwargs...` are passed to this function.
+
+# Optional Arguments
+* `throw_error`       - (`false`) throw an error message if the gradient is wrong
+
+"""
+function is_Hessian_linear(
+    M,
+    Hess_f,
+    p,
+    X=rand(M; vector_at=p),
+    Y=rand(M; vector_at=p),
+    a=randn(),
+    b=randn();
+    throw_error=false,
+    io=nothing,
+    kwargs...,
+)
+    Z1 = Hess_f(M, p, a * X + b * Y)
+    Z2 = a * Hess_f(M, p, X) + b * Hess_f(M, p, Y)
+    isapprox(M, p, Z1, Z2; kwargs...) && return true
+    n = norm(M, p, Z1 - Z2)
+    m = "Hess f seems to not be linear since Hess_f(p)[aX+bY] differs from aHess f(p)[X] + b*Hess f(p)[Y] by $(n).\n"
+    (io !== nothing) && print(io, m)
+    throw_error && throw(ErrorException(m))
+    return false
+end
+
+@doc raw"""
+    is_Hessian_symmetric(M, Hess_f, p=rand(M), X=rand(M; vector_at=p), Y=rand(M; vector_at=p);
+    throw_error=false, io=nothing, atol::Real=0, rtol::Real=atol>0 ? 0 : √eps
+)
+
+Check whether the Hessian function `Hess_f` fulfills symetry, i.e. that
+
+```math
+⟨\operatorname{Hess} f(p)[X], Y⟩ = ⟨X, \operatorname{Hess} f(p)[Y]⟩
+```
+
+which is checked using `isapptox` and the `kwargs...` are passed to this function.
+
+# Optional Arguments
+
+* `atol`, `rtol` - with the same defaults as the usual `isapprox`
+* `throw_error` - (`false`) throw an error message if the gradient is wrong
```

Comment:
```suggestion
* `throw_error` - (`false`) throw an error message if the Hessian is wrong
```

Reply:
Maybe a few things were copied to carelessly ;)